nmmo/task/task.py

```diff
@@ -0,0 +1,467 @@
+    g = npi.group_by(entity_all[:,ent_cols["population_id"]])
+    for pop_id, ents in zip(*g(entity_all[:,ent_cols["id"]])):
+      alive_byteam[int(pop_id)] = ents
+
+    item_all = ItemState.Query.table(realm.datastore).astype(np.int32)
+    item_cols = ItemState.State.attr_name_to_col
+
+    tile_cols = TileState.State.attr_name_to_col
+
+    # CHECK ME: what gets return for eliminated teams? killed agents?
+    for pop_id, spawn_pos in self.team2ent.items():
+      flt_ent = entity_all[:,ent_cols['population_id']] == pop_id
+      flt_item = np.isin(item_all[:,item_cols['owner_id']], list(spawn_pos.keys()))
+
+      team_gs[pop_id] = TeamGameState(
+        tick = realm.tick,
+        config = self.config,
+        pop_id = int(pop_id),
+        spawn_pos = spawn_pos,
+        alive_agents = alive_agents,
+        alive_byteam = alive_byteam,
+        env_obs = { ent_id: env_obs[ent_id]
+                    for ent_id in spawn_pos.keys() if ent_id in env_obs },
+        entity_cols = ent_cols,
+        entity_data = entity_all[flt_ent],
+        item_cols = item_cols,
+        item_data = item_all[flt_item],
+        tile_cols = tile_cols,
+        cache_result = {}
+      )
+
+    return team_gs
+
+  # TODO(kywch)
+  # most entity/item info can be retrieved from the datastore, but some won't.
+  # in that case, we need a simple dataclass to pass remaining info
+
+
+class PredicateTask:
+  def __init__(self, *args):
+    self.name = self._task_name(args)
+
+  def _task_name(self, args):
+    tmp_list = [self.__class__.__name__]
+    for arg in args:
+      if isinstance(arg, type): # class
+        # str(arg) gives something like:
+        # "<class 'nmmo.systems.item.Ration'>", "<class 'nmmo.io.action.Melee'>"
+        tmp_list.append(str(arg)[1:-2].rsplit('.', maxsplit=1)[-1])
+      elif "object at" in str(arg):
+        tmp_list.append(str(arg).split(' object', maxsplit=1)[0].split('.')[-1])
+      elif arg is None:
+        tmp_list.append('Any')
+      else:
+        tmp_list.append(str(arg))
+    return '_'.join(tmp_list).replace(' ', '')
+
+  # pylint: disable=inconsistent-return-statements
+  def __call__(self, team_gs: TeamGameState, ent_id: int) -> bool:
+    """One should describe the code how evaluation is done.
+       LLM wiil use it to produce goal embedding, which will be
+       used by the RL agent to produce action.
+    """
+    # base predicates will use super().evaluate(team_gs, ent_id) to check
+    # if the provided ent_id can access team_gs
+    assert team_gs.is_member(ent_id), \
+      "The entity must be in the team to access the provided team gs"
+
+    # check if the cached result is available, and if so return it
+    if self.name in team_gs.cache_result:
+      return team_gs.cache_result[self.name]
+
+  def _desc(self, class_type):
+    return {
+      "type": class_type,
+      "name": self.name,
+      "evaluate": self.__call__.__doc__
+    }
+
+  def description(self) -> Dict:
```

Comment:
This was to deal with AND, OR, NOT, IMPLY, where I appended child node info, like:
    `desc.update({ 'desc_child': ["AND"] + [t.description() for t in self._tasks] })`

My idea of the description() -> Dict function is to retain all the task information, so that we can turn those into LLM prompts, for goal embedding, etc.